### PR TITLE
slow_conv2d grad_weight: call gemm directly

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65759
* #65758
* #65757
* #65737
* __->__ #65726

This PR isn't strictly necessary since grad_weight doesn't use
parallel_for. However, this does reduce the function overhead and will
make it easier to parallelize in the future.

Differential Revision: [D31257877](https://our.internmc.facebook.com/intern/diff/D31257877)

### CODE REVIEW

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).

File: aten/src/ATen/native/ConvolutionMM2d.cpp [hunk 1]
```diff
@@ -366,18 +384,24 @@ static void slow_conv2d_backward_weight_out_cpu_template(
 
   auto input = input_.contiguous();
   auto grad_output = grad_output_.contiguous();
+  TORCH_CHECK(grad_weight.is_contiguous(), "slow_conv2d: grad_weight must be contiguous");
```
Comment (ngimel) [inline aten/src/ATen/native/ConvolutionMM2d.cpp:387]: this is checked earlier, in line 369

File: aten/src/ATen/native/ConvolutionMM2d.cpp [hunk 2]
```diff
@@ -319,14 +319,32 @@ void slow_conv2d_backward_out_cpu_template(
...
+    TensorAccessor<scalar_t, 3> grad_output,
+    TensorAccessor<scalar_t, 2> finput) {
+  // Compute grad_weight += grad_output.reshape({grad_output.shape(0), -1}) * finput.T
+  // Note gemm expects fortran order, so all 3 matrices are transposed.
+  // Swapping argument order cancels this, since C == AB <=> T(C) == T(B)T(A)
+  const int64_t m = finput.size(0);
+  const int64_t n = grad_output.size(0);
+  const int64_t k = grad_output.size(1) * grad_output.size(2);
+
+  constexpr int64_t one = 1;
```
Comment (ngimel) [inline aten/src/ATen/native/ConvolutionMM2d.cpp:334]: this is unused var (this is a merge-blocking error internally)

File: aten/src/ATen/native/ConvolutionMM2d.cpp [hunk 3]
```diff
@@ -366,18 +384,24 @@ static void slow_conv2d_backward_weight_out_cpu_template(
 
   auto input = input_.contiguous();
```
Comment (ngimel) [inline aten/src/ATen/native/ConvolutionMM2d.cpp:385]: this looks unnecessary, you only need `input.size(0)` and `input.scalar_type()` below, both can be called on `input_`.

File: aten/src/ATen/native/ConvolutionMM2d.cpp (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/ConvolutionMM2d.cpp:387]: You also need to check that finput is contiguous?

Comment (peterbell10) [reply aten/src/ATen/native/ConvolutionMM2d.cpp]: Changed to check `finput`.

File: aten/src/ATen/native/ConvolutionMM2d.cpp (hunk 2 above)
Comment (peterbell10) [reply aten/src/ATen/native/ConvolutionMM2d.cpp]: Ah, I added the fixup removing this line to the wrong commit in the stack. Fixed now.

File: aten/src/ATen/native/ConvolutionMM2d.cpp (hunk 3 above)
Comment (peterbell10) [reply aten/src/ATen/native/ConvolutionMM2d.cpp]: Agreed, removed.

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
Can you please rebase this stack? I'm getting the following when I try to import it
```
RuntimeError: We tried to import new changes for D31257876, but this diff has already been landed.  This could occur if you rebased your stack prior to the landed diff being pushed by ShipIt; in this case, wait until this diff is exported with ShipIt, rebase on master, and try again.
```

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257877).

**Pull request closed without merging**